### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
# having a look at the data
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [3]:
# calculating the number of unique players from the SN column
player_count = purchase_data['SN'].nunique()
player_count_df = pd.DataFrame({'Total Players': [player_count]})
# displaying the results: total number of players
player_count_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
# running the basic calculations for the Purchasing Analysis
unique_items = purchase_data['Item ID'].nunique()
avg_price = purchase_data['Price'].mean()
num_purchases = purchase_data["Purchase ID"].count()
total_rev = num_purchases * avg_price

In [5]:
# creating the summary data frame with the required format
purchasing_analysis_df = pd.DataFrame({"Number of Unique Items": [unique_items],
                                   "Average Price": "${:.2f}".format(avg_price),
                                   "Number of Purchases": num_purchases,
                                   "Total Revenue": "${:,.2f}".format(total_rev)})
# displaying the summary data frame
purchasing_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [6]:
# grouping by gender with 'SN' only and store in new data frame as nunique()
gender_grouped = purchase_data[['Gender', 'SN']].groupby(['Gender']).nunique()
# create an add the percentage calculated series with the pct format
gender_pct = gender_grouped["SN"] / player_count
gender_grouped['Percentage of Players'] = gender_pct.map("{:.2%}".format)
# keep the columns needed and rename 'SN'
gender_extract = gender_grouped.iloc[:, 1:]
gender_renamed = gender_extract.rename(columns = {"SN": "Total Count"})
# sort and change the number format of "Percentage of Players" to percentage
gender_demo_df = gender_renamed.sort_values('Total Count', ascending=False)
# removing the index header and displaying the results: gender demographics data frame
gender_demo_df.index.name = None
gender_demo_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# Grouping by Gender with the required colums for calculations
gender_pa_grouped = purchase_data[['Gender', 'Purchase ID', 'SN', 'Price']].groupby(['Gender'])
# running the basic calculations for the required analysis
purchase_count = gender_pa_grouped['Purchase ID'].nunique()
avg_purchase_price = gender_pa_grouped['Price'].mean()
total_purchase_val = gender_pa_grouped['Price'].sum()
total_count = gender_pa_grouped['SN'].nunique()
avg_purchas_pperson = total_purchase_val/total_count
# creating the summary data frame, purchase analysis by gender
purchase_analysis_by_gender_df = pd.DataFrame({"Purchase Count": purchase_count,
                                           "Average Pruchase Price": avg_purchase_price.map("${:,.2f}".format),
                                           "Total Purchase Value": total_purchase_val.map("${:,.2f}".format),
                                           "Avg Total Purchase per Person": avg_purchas_pperson.map("${:,.2f}".format)})
# displaying the results: purchase analysis by gender data frame
purchase_analysis_by_gender_df

Purchase Count Average Pruchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [8]:
# establishing the labels and bins for ages
label_group = ['<10', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']
bins = [0, 9.99, 14.99, 19.99, 24.99, 29.99, 34.99, 39.99, 100]
# creating a copy of the purchase_data data frame
pd_df = purchase_data[['Age', 'Gender', 'Purchase ID', 'SN', 'Price']]
# categorizing the existing players by using the age bins
pd_df['Age Bins'] = pd.cut(pd_df['Age'], bins, labels=label_group, include_lowest=True)
# grouping by age
age_grouped_df = pd_df.groupby(['Age Bins'])
# running the basic calculations for the required analysis
total_count_age = age_grouped_df['SN'].nunique()
pct_players_age = total_count_age/ player_count
# creating the summary data frame, age demographics
age_demo_df = pd.DataFrame({'Total Count': total_count_age,
                           'Percentage of Players': pct_players_age.map("{:.2%}".format)})
# removing index header and displaying the results: age demographics data frame
age_demo_df.index.name = None
age_demo_df

/Users/emiliobello/opt/anaconda3/envs/PythonData/lib/python3.6/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
# using the same group df created from the previous activity: age_grouped_df
# running the basic calculations for the required analysis
purchase_count_age = age_grouped_df['Purchase ID'].nunique()
avg_purchase_price_age = age_grouped_df['Price'].mean()
total_purchase_val_age = age_grouped_df['Price'].sum()
total_count_age = age_grouped_df['SN'].nunique()
avg_purchas_pperson_age = total_purchase_val_age/total_count_age
# creating the summary data frame, purchase analysis by age
purchase_analysis_by_age_df = pd.DataFrame({"Purchase Count": purchase_count_age,
                                           "Average Purchase Price": avg_purchase_price_age.map("${:,.2f}".format),
                                           "Total Purchase Value": total_purchase_val_age.map("${:,.2f}".format),
                                           "Avg Total Purchase per Person": avg_purchas_pperson_age.map("${:,.2f}".format)})
# adding index header and displaying the results: purchase analysis by age data frame
purchase_analysis_by_age_df.index.name = 'Age Ranges'
purchase_analysis_by_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:
# grouping by spender
sn_grouped_df = purchase_data[['SN','Purchase ID','Price']].groupby(['SN'])
# running the basic calculations for the required analysis
purchase_count_sn = sn_grouped_df['Purchase ID'].count()
avg_purchase_price_sn = sn_grouped_df['Price'].mean()
total_purchase_val_sn = sn_grouped_df['Price'].sum()
# creating the summary data frame, top spender
sn_df = pd.DataFrame({"Purchase Count": purchase_count_sn,
                      "Average Purchase Price": avg_purchase_price_sn,
                      "Total Purchase Value": total_purchase_val_sn})
# sorting descending and indicating the head(5), since it won't work after formatting
sn_sorted_df = sn_df.sort_values(["Total Purchase Value"], ascending=False).head()
# formatting and displaying the results: top spenders
sn_formatted_df = sn_sorted_df.style.format({"Purchase Count": "{:.0f}".format,
                                      "Average Purchase Price": "${:,.2f}".format,
                                      "Total Purchase Value": "${:,.2f}".format})
sn_formatted_df

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [11]:
# grouping by item
items_grouped_df = purchase_data[['Item ID', 'Item Name', 'Price']].groupby(['Item ID', 'Item Name'])
# running the basic calculations for the required analysis
purchase_count_items = items_grouped_df['Item ID'].count()
total_purchase_val_items = items_grouped_df['Price'].sum()
item_price_items = total_purchase_val_items/purchase_count_items
# creating the summary data frame, most popular item
items_df = pd.DataFrame({"Purchase Count": purchase_count_items,
                      "Item Price": item_price_items,
                      "Total Purchase Value": total_purchase_val_items})
# sorting descending and indicating the head(5), since it won't work after formatting. 
# We'll use this DF for the most profitable item sort by total purchase value
items_sorted_df = items_df.sort_values(["Purchase Count"], ascending=False).head()
# formatting and displaying the results: most popular items data frame
items_formatted_df = items_sorted_df.style.format({"Purchase Count": "{:.0f}".format,
                      "Item Price": "${:,.2f}".format,
                      "Total Purchase Value": "${:,.2f}".format})
items_formatted_df

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [12]:
# sorting by total purchase value the previous unformatted items_df
p_items_sorted_df = items_df.sort_values(["Total Purchase Value"], ascending=False).head()
# formatting and displaying the results: most profitable items data frame
p_items_formatted_df = p_items_sorted_df.style.format({"Purchase Count": "{:.0f}".format,
                      "Item Price": "${:,.2f}".format,
                      "Total Purchase Value": "${:,.2f}".format})
p_items_formatted_df

-------------------------------------------------------------------------------------------------------------------

# Observable Trends

- As observed on the Gender Demographics, males represent 84% of the Heroes of Pymoli players.
- Ages ranging from 20 to 24, represent the higher number of players of Heroes of Pymoli, as well as the total purchase value, followed by ages between 15 to 19.
- It is an observable trend that Heroes of Pymoli is more popular among male adolescents and young adults.
- Items with ID 92 and 178 figure in both the most popular and most profitable items.